database.cc

```diff
@@ -1136,7 +1136,8 @@ void keyspace::mark_as_populated() {
 
 
```

Comment:
> -  the data will be written using CL=ALL and read using CL=ONE; thanks
>    to this, restarting node won't have to communicate with other nodes
>    to retrieve the data of the last known generation. Note that writing
>    with CL=ALL does not reduce availability: creating a new generation
>    *requires* all nodes to be available anyway, because they must learn
>    about the generation before their clocks go past the generation's
>    timestamp; if they don't, partitions won't be mapped to stream IDs
>    consistently across the cluster

This means that if a node is down, we cannot perform a node addition or removal.

It was already iffy, since we don't yet enable @asias repair-based node operations (which do allow node additions and removal with nodes down).

This is a negative side effect of this series, but I think we can tolerate it because those operations were quite shaky before. Once we have raft both node operations and everywhere replication will be safer in the face of downed (or partitioned) nodes.

Reply:
> repair-based node operations (which do allow node additions and removal with nodes down).

Do they? Decommissioning a node when some nodes are down may cause data loss. Today decommissioning ensures that all data is moved to other nodes. Consider RF=1 table - you must move data from the leaving node to _every other node_ (assuming that every node owns at least one token range), so you cannot allow some nodes to be down during the operation. There is also `removenode` which is the unsafe counterpart of decommission (one which allows dropping data). And our docs explicitly warn from using `removenode`. (https://docs.scylladb.com/operating-scylla/procedures/cluster-management/remove_node/)

Similarly node bootstrap requires all nodes to be up (if we don't want data loss) because data must be streamed from _every_ node. Again consider RF=1 case.

CDC always assumed that all nodes are up when bootstrapping a node, the assumption does not change with this PR.

Reply:
Obviously RF=1 cannot be decommissioned with a node down (@asias how will repair-based-node-ops react?). But with RF=3, it will either move the data to the replica that will own the data (if it is up) or to two other replicas (if not), preserving quorum.

(there is a negative effect to this in that a large repair will be needed later).

Reply:
Even with RF=3 you could have a successful CL=ONE write which ended up on exactly one replica, and is supposed to be observable by all later successfull CL=ALL reads unless a data-loss event happens in between which forces the admin to perform an unsafe administrative operation such as `nodetool removenode` (which - as the documentation properly points out - does not preserve consistency of operations). `nodetool decommission` is supposed to be a **safe** operation, preserving consistency, in particular CL=ONE writes -> CL=ALL reads consistency ("read-your-writes"). If the admin accepts possibility of dropping consistency and really wants to remove a node when some nodes are down, they are supposed to use `nodetool removenode`.

Bootstrap is symmetric to `nodetool decommission` and is also supposed to be a completely safe operation, preserving consistency of operations and not losing data. For that it **must** require all other nodes to be up in general.

I see no reason why we would want to lower the requirements for bootstrap. If some nodes are irrecoverably down, the user should replace or remove them and bring a cluster to a total UN state before bootstrapping a new node.

Reply:
CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.

The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.

Reply:
> CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.

I'm considering a scenario where no replica is lost. Should we drop our promises in that case as well?

The very true generalization of your statement is: for any RF, CL=(whatever) write + CL=(whatever) read is not a safe operation, because you could lose RF replicas before you replace any of them (doesn't matter if you repair or not).

That's right, but **if no disaster happens**, then we should guarantee that CL=QUORUM reads observe CL=QUORUM writes. Similarly, we should guarantee that CL=ALL reads observe CL=ONE writes.

In fact it's common knowledge in the Cassandra community - and I'm guessing Scylla community too, since people assume they give similar guarantees - that in order to read back previously written data, the following must be satisfied: `write CL + read CL > RF`. For example: https://stackoverflow.com/questions/25043208/read-operation-in-cassandra-at-consistency-level-of-quorum. As the accepted answer says, examples are: `write ALL + read ONE`, `write ONE + read ALL`, `write QUORUM + read QUORUM`. When I first heard about Scylla on a lecture by @haaawk at my university, that's what was written on the presentation too.

> The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.

Ok, but bootstrap should be safe (not allow consistency loss) by default. Bootstrapping in presence of downed nodes should be an override config option and heavily discouraged in docs (reserved only for disaster scenarios). The override should contain a promise from the administration that the downed node is lost permanently. Given such a promise we can also safely introduce a new CDC generation. The CL=ALL write in this table would be then changed to CL=ALL-ONE write.

Reply:
> > CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.
> 
> I'm considering a scenario where no replica is lost. Should we drop our promises in that case as well?
> 

We should see if those promises are worth keeping. CL=ALL means that as soon as you perform maintenance of have a crash you lose availability. Write(CL=ONE) + read(CL=ALL) gives really good availability for write, but poor durability. More reasonable is write(CL=ALL) + read(CL=ONE). Compromising maintainability for these uncommon scenarios is not a good tradeoff IMO.

> The very true generalization of your statement is: for any RF, CL=(whatever) write + CL=(whatever) read is not a safe operation, because you could lose RF replicas before you replace any of them (doesn't matter if you repair or not).

RF=3 only handles one node failure. You must handle this node failure soon; until you do, another failure can cause you data loss. If you want to survive two node failures, you must use RF=5.

> 
> That's right, but **if no disaster happens**, then we should guarantee that CL=QUORUM reads observe CL=QUORUM writes. Similarly, we should guarantee that CL=ALL reads observe CL=ONE writes.

QUORUM/QUORUM safety is maintained with repair-based node operations. Write(ONE)/Read(ALL) might not be, if the node is downed before it had a chance to replicate, but write(ONE) implicitly gives up durability because a disaster *might* happen. Write(ALL)/Read(ONE) is still safe.

> 
> In fact it's common knowledge in the Cassandra community - and I'm guessing Scylla community too, since people assume they give similar guarantees - that in order to read back previously written data, the following must be satisfied: `write CL + read CL > RF`. For example: https://stackoverflow.com/questions/25043208/read-operation-in-cassandra-at-consistency-level-of-quorum. As the accepted answer says, examples are: `write ALL + read ONE`, `write ONE + read ALL`, `write QUORUM + read QUORUM`. When I first heard about Scylla on a lecture by @haaawk at my university, that's what was written on the presentation too.

Well, if permanent node failures are not ruled out then write(ONE)+read(ALL) will not work, either in Cassandra or Scylla.

> 
> > The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.
> 
> Ok, but bootstrap should be safe (not allow consistency loss) by default. Bootstrapping in presence of downed nodes should be an override config option and heavily discouraged in docs (reserved only for disaster scenarios). The override should contain a promise from the administration that the downed node is lost permanently. Given such a promise we can also safely introduce a new CDC generation. The CL=ALL write in this table would be then changed to CL=ALL-ONE write.

It is too strict. The downed node may be recovered, and repair-based node operations will ensure data safety. W1RA is too rare to align everything with its requirements.

For CDC, we should use Raft instead of these heuristics.

Reply:
> We should see if those promises are worth keeping. CL=ALL means that as soon as you perform maintenance of have a crash you lose availability. Write(CL=ONE) + read(CL=ALL) gives really good availability for write, but poor durability. More reasonable is write(CL=ALL) + read(CL=ONE). Compromising maintainability for these uncommon scenarios is not a good tradeoff IMO.

W+R>RF is a simple to remember and understand rule, and users expect it to work intuitively; e.g. if you wrote data to one replica and read from all, you contact that one replica, so you see the data. Introducing exceptions to this rule makes systems built on top of Scylla/Cassandra harder to reason about and understand.

Another path we could take while keeping simple rules is to drop all CLs other than QUORUM altogether, high availability and consistency, and tolerance of RF/2 failures. AFAIR DynamoDB took this approach.

> RF=3 only handles one node failure.

I think you implicitly assume quorum operations. Under some other strategies it can handle 2 failures. For example consider the following strategy. Users always read with CL=1, and:
- until 3 nodes are available, write with CL=3
- after one failure, write with CL=2
- after second failure, write with CL=1

With this strategy we remain consistent and available and tolerate up to 2 failures.

> but write(ONE) implicitly gives up durability because a disaster _might_ happen

"write(TWO) implicitly gives up durability because 2 disasters _might_ happen" (correlated failure).

I don't think it's a question of what _might_ happen. Write(ONE) tolerates 0 failures, yes. But that doesn't mean Scylla should be the one to actually introduce the failure. Allowing bootstrap with a partitioned or temporarily shut down node is introducing a failure.

> Well, if permanent node failures are not ruled out then write(ONE)+read(ALL) will not work, either in Cassandra or Scylla.

And what should we guarantee **before** the failure happens?

> It is too strict. The downed node may be recovered, and repair-based node operations will ensure data safety. 

So even if  the down node is recoverable we still want to allow bootstrap to proceed, just so the entire procedure is a bit faster? I don't see much value here, considering the rarity of those operations. You're considering a situation where a node permanently fails and then very soon after the admin decides that it's a good moment to scale up. I think that when nodes permanently fail, admins' first priority is to stabilize the cluster (replacing the failed nodes), not scaling up.

> For CDC, we should use Raft instead of these heuristics.

Raft does not automatically solve every problem. Maybe it would if everything would go through a single state machine replicated by a single Raft group, but such solutions don't scale. Thus you need communication between independent Raft state machines, so you must solve problems "outside Raft", and CDC generation management is full of problems that must be solved outside Raft.

That said, when we introduce **Raft-driven tablets**, then it will be possible to safely introduce a new CDC generation without all nodes being up - only the leader of each tablet will have to be available.

Reply:
I have been using the "CL=ONE write + CL=ALL read" example to convince Avi that we should reject the bootstrap by default if a node is down during bootstrap instead of stream from quorum nodes and continue. 

In case the node is down temporarily, we could avoid data loss if we reject the bootstrap ops.

In case the node is really down and not way to recover, or there is no enough time to recover the node, or the user does not bother to recover the node and think it is more important to add a new node immediately than losing the data, we can use the ignore_nodes to exclude the down nodes explicitly like we do in the latest removenode and replacing and regular repair operation.

It is much easier for user to reason about the system:  we are on the safe side by default not matter what, if you want to bootstrap  with the node down you might lose data, users have to ask for it explicitly. 

There are other reason we need all nodes to be up during bootstrap. A node might be partitioned from the bootstrap node, it is not down and can server user reads /writes, if we continue the bootstrap, the partitioned node will use wrong topology which can cause data consistency issue.  Again, to be safe by default, we should reject the bootstrap in such case.

Reply:
> > For CDC, we should use Raft instead of these heuristics.
> 
> Raft does not automatically solve every problem. Maybe it would if everything would go through a single state machine replicated by a single Raft group, but such solutions don't scale. Thus you need communication between independent Raft state machines, so you must solve problems "outside Raft", and CDC generation management is full of problems that must be solved outside Raft.

It is the same for topology change. People think raft will magically allow us to do topology change with some nodes down since we can achieve some quorum using raft. Raft will not help as long as we still use distributed tokens to form the token ring for read / write. We need all nodes to be aware of this change to maintain consistency.

Reply:
With Raft, a node that is partitioned from the rest of the cluster will be aware that it is partitioned and will not break consistency. It can be done this way:

1. Raft creates a "topology version number" each time topology changes
2. A coordinator includes the topology version number in each request
3. Replicas reject requests that have mismatched topology version number

Note that each topology change generates two versions. Let's assume we start from version v0.
1. Node is added, we start double-writes. Generate v1 and tell nodes to accept v1 and v0.
2. All live nodes are aware of v1. Tell them to reject v0 and wait for ack.
3. Start streaming data (maybe via repair).
4. Streaming is complete, generate v2 that has the new topology without double writes (but accept v1).
5. All live nodes are aware of v2, Tell them to reject v1 and execute cleanup.

Reply:
Consider a cluster of 9 nodes and a keyspace with RF = 3. Suppose there are 3 "tablets" / "replica sets" / "Raft groups" (however you want to name them) replicating different parts of the keyspace: {n1, n2, n3}, {n4, n5, n6}, {n7, n8, n9} (there may be other unlisted tablets, e.g. {n1, n4, n7}, but we don't care for this discussion). Suppose that n1 is the leader/coordinator of the first group, n4 of the second, n7 of the third. Suppose that the third group is partitioned from the rest. Recall that all nodes are also members of Raft Group 0 which handles topology. Observe that {n1, n2, n3, n4, n5, n6} form a quorum in Raft Group 0, thus they can reach a decision on changing the topology of the cluster. Unfortunately, even though the topology was changed, the group {n7, n8, n9} does not know about it because it was partitioned. However, the entire group is running and all nodes within the group can contact each other, so the leader of the group - n7 - can still perform writes, unaware of the topology change. It keeps using the same "topology version number".